README
Retchmail is intended as a blazingly fast replacement for ESR's Fetchmail.
For a much more in depth overview of the philosophy behind this program,
please read apenwarr's ANNOUNCEment for the program, which should be
found in the same directory as this file.

For a quick overview of the functionality, please see the retchmail(1) man
page, which should get you started.

Questions, comments, feature requests, and especially patches are welcome.
Please see the Retchmail web page for more information:

http://alumnit.ca/wiki/index.php?page=RetchMail

This program is Copyright 2002-2003 Net Integration Technologies Inc.

This program is licensed under the LGPL, a full copy of which can be found
in the COPYING.LIB file in the same directory as this file.


debian/changelog
retchmail (1:1.1-1) unstable; urgency=low

  * New upstream version.  (Closes: #407873)

 -- William Lachance <wrlach@gmail.com>  Thu, 19 Jul 2007 19:30:09 -0400

retchmail (1:1.0-1) unstable; urgency=low

  * New upstream version.

 -- Simon Law <sfllaw@debian.org>  Tue, 22 Nov 2005 20:58:10 -0500

retchmail (1:0.09-3) unstable; urgency=low

  * Update config documentation. (Closes: #302884)

 -- James Morrison <phython@debian.org>  Sun,  3 Apr 2005 13:19:30 -0400

retchmail (1:0.09-2) unstable; urgency=low

  * Only use low bits of WvLog pointer in WvHash function. (Closes: #302372)
  * Use xplc 0.3.11.

 -- James Morrison <phython@debian.org>  Sat,  2 Apr 2005 11:23:20 -0500

retchmail (1:0.09-1) unstable; urgency=low

  * New upstream release.  (Closes: #290999)

 -- James Morrison <phython@debian.org>  Wed, 22 Mar 2004 22:16:03 -0500

retchmail (1:0.04-2) unstable; urgency=low

  * Don't call callbacks when they are empty. (Closes: #232469)

 -- James Morrison <phython@debian.org>  Wed,  3 Mar 2004 17:16:03 -0500

retchmail (1:0.04-1) unstable; urgency=low

  * New upstream release.

 -- James A. Morrison <phython@debian.org>  Sun, 15 Feb 2004 13:34:48 -0500

retchmail (1:0.03+20031110-1) unstable; urgency=low

  * Fix Build-Depends.

 -- James Morrison <phython@debian.org>  Mon, 10 Nov 2003 10:17:17 -0500

retchmail (1:0.03+20031017-1) unstable; urgency=low

  * Synchronize with upstream CVS.

 -- James A. Morrison <phython@debian.org>  Fri, 17 Oct 2003 17:48:36 -0400

retchmail (0.3) unstable; urgency=high

  * Synchronize with libwvstreams 3.71

  * Make it so that we handle email addresses with @ in them

  * Handle APOP thanks to a patch from markj@cloaked.freeserve.co.uk

  * Fixes core-dump on certain messages

  * Debian Packaging Fixes (Closes: #142527, #142567)

 -- Patrick Patterson <ppatters@nit.ca>  Mon, 29 Jul 2002 18:39:51 -0400

retchmail (0.2) unstable; urgency=low

  * Synchronise with libwvstreams

  * Debian Packaging Fixes

  * Initial Debian release (Closes: #138948)

 -- Patrick Patterson <ppatters@debian.org>  Mon, 28 Jan 2002 14:01:04 -0500

retchmail (0.1) unstable; urgency=low

  * Initial Release.

 -- Patrick Patterson <ppatters@nit.ca>  Mon, 28 Jan 2002 14:01:04 -0500



retchmail.cc
/*
 * Worldvisions Weaver Software:
 *   Copyright (C) 1997-2005 Net Integration Technologies, Inc.
 *
 * Avery's insanely fast alternative to Fetchmail
 *
 * This code is LGPL - see the file COPYING.LIB for a full copy of the
 * license.
 */

// WvStreams headers
#include "strutils.h"
#include "uniconfroot.h"
#include "wvlockfile.h"
#include "wvver.h"
#include "wvtcp.h"
#include "wvlog.h"
#include "wvpipe.h"
#include "wvistreamlist.h"
#include "wvlogrcv.h"
#include "wvsslstream.h"
#include "wvhashtable.h"
#include "wvargs.h"
#include "wvcrash.h"
#include "wvhex.h"

// Retchmail headers
#include "wvsendmail.h"
#include "wvpopclient.h"

#include <signal.h>
#include <assert.h>
#include <pwd.h>
#include <sys/types.h>

struct LogNum
{
    int num;
    WvString src;
};

DeclareWvDict(LogNum, WvString, src);


class RetchLog : public WvLogConsole
{
public:
    RetchLog(WvLog::LogLevel _max_level) 
	: WvLogConsole(1, _max_level), lognums(5)
	{ maxnum = 0; }
    virtual ~RetchLog() { }
    
    virtual void _begin_line();
    
    int maxnum;
    LogNumDict lognums;
};


void RetchLog::_begin_line()
{
    if (!strncmp(last_source, "PopRetriever ", 13))
    {
	LogNum *lognum = lognums[last_source];
	if (!lognum)
	{
	    lognum = new LogNum;
	    lognum->num = maxnum++;
	    lognum->src = last_source;
	    lognums.add(lognum, true);
	}
	
	// identify the connection without being too verbose
	if (lognum->num < 26)
	{
	    char str[2];
	    str[0] = 'A' + lognum->num;
	    str[1] = 0;
	    write(str);
	}
	else
	    write(lognum->num);
	if (last_level <= WvLog::Info)
	    write("* ");
	else
	    write("  ");
    }
    else
	WvLogConsole::_begin_line(); // just print the whole thing
}



static volatile bool want_to_die = false;
void signal_handler(int signum)
{
    fprintf(stderr, "\nCaught signal %d; cleaning up and terminating.\n",
	    signum);
    want_to_die = true;
    signal(signum, SIG_DFL);
}


static WvPopClient *newpop(WvStringParm acct,
			   WvStringParm _pass, WvStringParm _deliverto,
                           WvStringParm _mda, bool flush, bool apop_en,
                           bool apop_fall_en, bool explode, bool safemode,
			   bool ignorerp)

{
    WvString user(acct), serv, pass(_pass), deliverto(_deliverto),
            mda(_mda);
    bool ssl = false;
    
    char *cptr = strrchr(user.edit(), '@');
    if (cptr)
    {
	*cptr++ = 0;
	serv = cptr;
    }
    else
	serv = "localhost";
    
    cptr = strchr(serv.edit(), ':');
    if (!cptr)
    {
	serv.append(":110");
	cptr = strchr(serv.edit(), ':'); // guaranteed to work now!
    }
    
    if (atoi(cptr+1) == 995)
	ssl = true;
    
    WvTCPConn *tcp = new WvTCPConn(serv);
    WvStream *conn = tcp;
    if (ssl) // FIXME: ssl verify should probably be set to something.
	conn = new WvSSLStream(tcp, NULL);
    
    return new WvPopClient(conn, acct, pass, deliverto, mda, flush, apop_en, apop_fall_en, explode, safemode, ignorerp);
}


static bool dec_log_level_cb(void *userdata)
{
    WvLog::LogLevel &level = *static_cast<WvLog::LogLevel *>(userdata);
    if ((int)level > (int)WvLog::Critical)
	level = (WvLog::LogLevel)((int)level - 1);
    return true;
}


static bool inc_log_level_cb(void *userdata)
{
    WvLog::LogLevel &level = *static_cast<WvLog::LogLevel *>(userdata);
    if ((int)level < (int)WvLog::Debug5)
	level = (WvLog::LogLevel)((int)level + 1);
    return true;
}


int main(int argc, char **argv)
{
    bool flush = false, explode = false, single_stream = false;
    WvLog::LogLevel lvl = WvLog::Debug1;
    WvString deliverto("");
    WvString confmoniker("");
    
    // Set up WvCrash
    wvcrash_setup(argv[0]);
    
    // make sure electric fence works
    free(malloc(1));

    signal(SIGPIPE, SIG_IGN);
    
    struct passwd *pw = getpwuid(getuid());
    if (pw)
	deliverto = pw->pw_name;

    WvArgs args;

    args.add_option('q', "quiet",
		    "Decrease log level (can be used multiple times)",
		    dec_log_level_cb, &lvl);
    
    args.add_option('d', "debug",
		    "Increase log level (can be used multiple times)",
		    inc_log_level_cb, &lvl);

    args.add_set_bool_option('F', "flush", 
			     "Flush (delete) messages after downloading",
			     flush);

    args.add_set_bool_option('E', "explode", 
			     "Send mail to the user on the system corresponding to the user the mail is sent to",
			     explode);
    
    args.add_option('t', "to", 
		    WvString("Deliver mail to <deliverto> (default '%s')", deliverto),
		    "deliverto", deliverto);

    args.add_option('c', "config",
		    "Use <moniker> instead of ini:~/.retchmail/retchmail",
		    "moniker", confmoniker);
    
    args.set_version("Retchmail version " RETCHMAIL_VER_STRING);
    args.set_email("<wvstreams-dev@lists.nit.ca>");

    args.add_set_bool_option('s', "singlestream", 
			     "Use only one TCP stream",
			     single_stream);
    
    WvStringList arguments;
    if (!args.process(argc, argv, &arguments))
	return 255;
    
    WvString lockname("/tmp/retchmail.%s.pid", getlogin());
    WvLockFile lockfile(lockname);

    if (!lockfile.lock())
    {
        if (lockfile.readpid() == -1)
            fprintf(stderr, "retchmail: can't access lockfile at %s.\n",
		    lockname.cstr());
        else
            fprintf(stderr, "retchmail: already running (%i).\n",
		    lockfile.readpid());
        return 3;
    }

    if (!deliverto)
    {
	fprintf(stderr, "retchmail: can't get username for uid#%u.  "
		"You must give the -t option.\n", getuid());
	return 3;
    }
    
    RetchLog logrcv(lvl);
    
    if (!confmoniker)
    {
	// attempting to use $HOME/.retchmail/retchmail.conf\n");
        confmoniker = "ini:";
        confmoniker.append(getenv("HOME"));
        confmoniker.append("/.retchmail/retchmail.conf");
    }

    UniConfRoot cfg(confmoniker);

    if (!cfg.haschildren())
    {
        wvcon->print("No data found, aborting\n");
	exit(1);
    }

    WvPopClient *cli = NULL;
    bool apop_enable = cfg["retchmail"]["Enable APOP"].getmeint(0);
    bool apop_enable_fallback = cfg["retchmail"]["Enable APOP Fallback"].getmeint(0);
    bool safemode = cfg["retchmail"]["Safe Mode"].getmeint(0);
    bool ignorerp = cfg["retchmail"]["Ignore Return-Path"].getmeint(0);
  
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);

    if (arguments.isempty())	    
    {
        UniConf sect = cfg["POP Servers"];
	if (sect.haschildren())
	{
	    UniConf::Iter i(sect);
	    for (i.rewind(); i.next(); )
	    {
		cli = newpop(i->key(), i->getme(),
			     cfg["POP Targets"][i->key()].getme(deliverto),
                             cfg["MDA Override"][i->key()].getme(
				     "/usr/sbin/sendmail"),
                             flush, apop_enable, apop_enable_fallback,
			     explode ? : cfg["Explode"][i->key()].getmeint(0),
			     safemode, ignorerp);
		WvIStreamList::globallist.append(cli, true, "client");
                if (single_stream)
                    while (!WvIStreamList::globallist.isempty() && !want_to_die)
    	                WvIStreamList::globallist.runonce();
	    }
	}
	else
	{
	    fprintf(stderr, "\n-- No config file and no accounts given on the "
		    "command line!\n");
	    args.print_usage(argc, argv);
	}
    }
    else
    {
	while (!arguments.isempty());
	{
	    WvString user(arguments.popstr());
	    WvString pass = cfg["POP Servers"][user].getme();
	    if (!pass)
	    {
		wvcon->print("Password for <%s>: ", user);
		
		system("stty -echo 2>/dev/null");
		pass = wvcon->blocking_getline(-1);
		system("stty echo 2>/dev/null");
		
		wvcon->print("\n");
	    }
	    
	    cli = newpop(user, pass,
			 cfg["POP Targets"][user].getme(deliverto),
			 cfg["MDA Override"][user].getme(
				 "/usr/sbin/sendmail"),
			 flush, apop_enable, apop_enable_fallback, explode,
			 safemode, ignorerp);
	    WvIStreamList::globallist.append(cli, true, "client");
            if (single_stream)
                while (!WvIStreamList::globallist.isempty() && !want_to_die)
    	            WvIStreamList::globallist.runonce();
	}
    }
    
    if (!single_stream)
        while (!WvIStreamList::globallist.isempty() && !want_to_die)
    	    WvIStreamList::globallist.runonce();

    lockfile.unlock();
}


wvpopclient.h
/*
 * Worldvisions Weaver Software:
 *   Copyright (C) 1997-2003 Net Integration Technologies, Inc.
 *
 * Avery's insanely fast alternative to Fetchmail -- Pop3 Client
 *
 * This code is LGPL - see the file COPYING.LIB for a full copy of the
 * license.
 */

#include "wvsendmail.h"
#include "wvstreamclone.h"
#include "wvstring.h"
#include "wvhashtable.h"
#include "wvlog.h"
#include "wvvector.h"

#ifndef WVPOPCLIENT_H
#define WVPOPCLIENT_H 1

DeclareWvDict(WvSendmailProc, int, count);

class WvPopClient : public WvStreamClone
{
public:
    WvSendmailProcDict sendprocs;
    WvString username, password, deliverto, mda;
    WvLog log;
    long res1, res2;
    int  next_req, next_ack, sendmails;
    bool flushing, apop_enable, apop_enable_fallback, explode;
    WvStringList trace;

    bool safemode;
    int max_requests;
    WvStringList safe_deletes;

    bool ignorerp;
    
    struct MsgInfo
    {
	int num;                // message number
	long len;               // message length (bytes)
	bool sent,              // message _fully_ transferred to sendmail
	     deleted;           // server acknowledged DELE command
	WvString err;		// error message
	int deletes_after_this; // number of DELE messages following this RETR

	MsgInfo()
	  : num(0), len(0), sent(false), deleted(false),
	    deletes_after_this(0)
	{
	}
    };
    DeclareWvVector(MsgInfo);
    MsgInfoVector mess;

    // note: we take possession of 'conn' and may delete it at any time!
    WvPopClient(WvStream *conn,
		WvStringParm acct, WvStringParm _password,
		WvStringParm _deliverto, WvStringParm _mda, 
		bool _flushing, bool _apop_enable,
                bool _apop_enable_fallback, bool _explode,
		bool _safemode, bool _ignorerp);
    virtual ~WvPopClient();

    bool never_select;
    virtual void pre_select(SelectInfo &si);
    virtual bool post_select(SelectInfo &si);
    virtual void execute();
    
    void cmd(WvStringParm s);
    void cmd(WVSTRING_FORMAT_DECL)
	{ return cmd(WvString(WVSTRING_FORMAT_CALL)); }
    bool response();
    
    void send_done(int count, bool success);

private:
    WvString acctparse(WvStringParm acct);
    bool not_found;
};

#endif // WVPOPCLIENT_H


wvsendmail.cc
#include "wvsendmail.h"

int WvSendmailProc::num_sendmails = 0;

WvSendmailProc::WvSendmailProc(const char **argv, int _count,
			       const WvSendmailCallback &_cb)
    : WvPipe(argv[0], argv, true, false, false), count(_count), cb(_cb)
{
    is_done = exited = false;
    num_sendmails++;
}


WvSendmailProc::~WvSendmailProc()
{
    close();
    
    int ret = finish();

    // be certain the callback gets called
    if (!exited && cb)
	cb(count, (ret == 0));
    
    num_sendmails--;
}


void WvSendmailProc::pre_select(SelectInfo &si)
{
    if (is_done && !exited && si.msec_timeout > 20)
	si.msec_timeout = 20;
    
    if (child_exited() && !exited && is_done)
	si.msec_timeout = 0;
    
    // another hack because isok() returns true when it shouldn't really
    if ((exited || is_done) && si.wants.writable)
	si.msec_timeout = 0;
    
    WvPipe::pre_select(si);
}


bool WvSendmailProc::post_select(SelectInfo &si)
{
    bool must = false;

    if (child_exited() && !exited && is_done)
    {
	exited = true;
	
	// call the callback
	if (cb)
	    cb(count, !exit_status());
	
	must = true;
    }
    
    // another hack because isok() returns true when it shouldn't really
    if ((exited || is_done) && si.wants.writable)
	must = true;
    
    return must || WvPipe::post_select(si);
}


bool WvSendmailProc::isok() const
{
    // note: this means people will try to write to us even if the pipe
    // says it's invalid!
    return WvPipe::isok() || !exited;
}


size_t WvSendmailProc::uwrite(const void *buf, size_t count)
{
    if (child_exited())
	return count; // fake it, because isok() is also faking it
    return WvPipe::uwrite(buf, count);
}


void WvSendmailProc::execute()
{
    WvPipe::execute();
    
    if (is_done && select(0, false, true, false))
	close();
}


void WvSendmailProc::done()
{
    is_done = true;
    force_select(false, true);
}


wvpopclient.cc
/*
 * Worldvisions Weaver Software:
 *   Copyright (C) 1997-2003 Net Integration Technologies, Inc.
 *
 * Avery's insanely fast alternative to Fetchmail -- Pop3 Client
 *
 * This code is LGPL - see the file COPYING.LIB for a full copy of the
 * license.
 */

#include <unistd.h>
#include "wvstring.h"
#include "wvpopclient.h"
#include "wvistreamlist.h"
#include "wvdigest.h"
#include "wvhex.h"
#include "strutils.h"

#define MAX_PROCESSES 5
#define MAX_REQUESTS 10


WvPopClient::WvPopClient(WvStream *conn, WvStringParm acct, 
			 WvStringParm _password,
			 WvStringParm _deliverto, WvStringParm _mda, 
                         bool _flushing, bool _apop_enable,
                         bool _apop_enable_fallback, bool _explode,
			 bool _safemode, bool _ignorerp)
  : WvStreamClone(conn), sendprocs(10),
    username(acctparse(acct)), // I hate constructors!
    password(_password), deliverto(_deliverto), mda(_mda),
    log(WvString("PopRetriever %s", acct), WvLog::Debug3),
    not_found(false)
{
    uses_continue_select = true;
    personal_stack_size = 65536;
    never_select = false;
    flushing = _flushing;
    apop_enable = _apop_enable;
    apop_enable_fallback = _apop_enable_fallback;
    next_req = next_ack = sendmails = 0;
    explode = _explode;
    safemode = _safemode;
    ignorerp = _ignorerp;
    max_requests = MAX_REQUESTS;

    if (safemode)
	max_requests = 1;
    safe_deletes.zap();
    
    log(WvLog::Info, "Retrieve mail from %s into %s.\n", acct, deliverto);
}


WvPopClient::~WvPopClient()
{
    if (isok()) // someone is killing us before our time
	seterr("Closed connection at user request.");
    
    terminate_continue_select();

    // disable callbacks for all remaining sendmail procs
    WvSendmailProcDict::Iter i(sendprocs);
    for (i.rewind(); i.next(); )
	i->cb = WvSendmailCallback();
    
    if (geterr())
	log(WvLog::Error, "Aborted.  Error was: %s\n", errstr());
    else
	log(WvLog::Info, "Done.\n");

    safe_deletes.zap();
}


WvString WvPopClient::acctparse(WvStringParm acct)
{
    WvString u(acct);
    char *cptr = strrchr(u.edit(), '@');
    
    if (cptr)
	*cptr = 0;
    return u;
}


void WvPopClient::cmd(WvStringParm s)
{
    // If we're already dead, we should try doing anything...
    if (!isok())
	return;
    
    print("%s\r\n", s);
    
    if (!strncasecmp(s, "pass ", 5))
	trace.append(new WvString("pass"), true);
    else if (!strncasecmp(s, "apop ", 5))
        trace.append(new WvString("apop"), true);
    else
    { 
        if (!strncasecmp(s, "apop ", 5))
            trace.append(new WvString("apop"), true);
        else
	    trace.append(new WvString(s), true);
    }
    
    if (!strncasecmp(s, "retr ", 5)) // less important
	log(WvLog::Debug3, "Send: %s\n", s);
    else if (!strncasecmp(s, "pass ", 5))    // hide passwords
	log(WvLog::Debug2, "Send: pass <NOT SHOWN>\n");
    else
	log(WvLog::Debug2, "Send: %s\n", s);
}


bool WvPopClient::response()
{
    flush(0);

    if (!isok())
    {
	log(WvLog::Critical,"Bailing out since the connection died: %s\n", 
	    errstr());
	return false;
    }
    
    assert(!trace.isempty());
    WvStringList::Iter i(trace);
    i.rewind(); i.next();
    
    char *line = blocking_getline(60*1000);
    if (!line)
    {
	if (isok())
	    seterr(ETIMEDOUT);
	return false;
    }
    
    log(WvLog::Debug2, "Recv(%s): %s\n", *i, line);
    i.unlink();
    
    if (!strncmp(line, "+OK", 3))
    {
	res1 = res2 = 0; 
	sscanf(line, "+OK %ld %ld", &res1, &res2);
	return true;
    }
    
    return false;
}


static int messcompare(const WvPopClient::MsgInfo *a,
		       const WvPopClient::MsgInfo *b)
{
    long base = 10240;
    long alen = a->len / base, blen = b->len / base;
    return (alen*base + a->num) - (blen*base + b->num);
}


void WvPopClient::pre_select(SelectInfo &si)
{
    bool oldrd = si.wants.readable;

    if (never_select)
	si.wants.readable = false;
    WvStreamClone::pre_select(si);
    si.wants.readable = oldrd;
}


bool WvPopClient::post_select(SelectInfo &si)
{
    bool val, oldrd = si.wants.readable;

    if (never_select)
	si.wants.readable = false;
    val = WvStreamClone::post_select(si);
    si.wants.readable = oldrd;
    return val;
}


void WvPopClient::execute()
{
    const char format[] = "%20.20s> %-40.40s\n";
    char *line, *greeting, *start, *end, *cptr;
    int count, nmsgs;
    WvString from, subj;
    bool printed, in_head, msgdone;
    
    WvStreamClone::execute();
    
    // read hello header
    // Adapted for APOP by markj@luminas.co.uk
    // trace.append(new WvString("HELLO"), true);
    greeting = blocking_getline(60*1000);
    if (!greeting || strncmp(greeting,"+OK", 3)) goto fail;
    // if (!response()) goto fail;
    log(WvLog::Debug1, "Recv(HELLO): %s", greeting);

    // log in
    if (apop_enable && strrchr(greeting, '>')) 
    {
        // APOP login attempt -- early code from fetchmail
        /* build MD5 digest from greeting timestamp + password */
        /* find start of timestamp */
        for (start = greeting;  *start != 0 && *start != '<';  start++)
            continue;
        if (*start == 0) goto fail;

        /* find end of timestamp */
        for (end = start;  *end != 0  && *end != '>';  end++)
            continue;
        if (*end == 0 || end == start + 1) goto fail;
        else
            *++end = '\0';
        // end of fetchmail code
	// 
        log(WvLog::Debug2, "Using APOP seed: %s\n", start);

        /* copy timestamp and password into digestion buffer */
        WvString digestsecret("%s%s",start,password);
        WvDynBuf md5buf;
        WvMD5Digest().flushstrbuf(digestsecret, md5buf, true /*finish*/);
        WvString md5hex = WvHexEncoder().strflushbuf(md5buf, true);
        
        log(WvLog::Debug2, "Using APOP response: %s\n", md5hex);
        cmd("apop %s %s", username, md5hex);
    } 

    if (!apop_enable || !strrchr(greeting, '>') 
	|| (!response() && apop_enable_fallback))
    {
        // USER/PASS login
        // some providers (i.e.: gmail) get unhappy if you send the user+pass
        // together, so wait for a response to the user cmd before sending
        // password.
        cmd("user %s", username);
        if (!response()) goto fail;
        cmd("pass %s", password);
        if (!response())
        {
           seterr("Server denied access.  Wrong password?");
           return;
        }
    }

    // get the number of messages
    cmd("stat");
    cmd("list"); // for later
    if (!response()) goto fail;
    log(WvLog::Info, "%s %s in %s bytes.\n", res1,
	res1==1 ? "message" : "messages", res2);
    nmsgs = res1;
    mess.set_capacity(nmsgs);
    
    // get the list of messages and their sizes
    if (!response()) goto fail;
    for (count = 0; count < nmsgs; count++)
    {
	line = blocking_getline(60*1000);
	mess.append(new MsgInfo(), true);
	if (!isok() || !line || 
	    sscanf(line, "%d %ld",
		   &(mess[count]->num), &(mess[count]->len)) != 2)
	{
	    log(WvLog::Error, "Invalid LIST response: '%s'\n", line);
	    goto fail;
	}
    }
    line = blocking_getline(60*1000);
    if (!isok() || !line || strcmp(line, ".") && strcmp(line, ".\r"))
    {
	log(WvLog::Error, "Invalid LIST terminator: '%s'\n", line);
	goto fail;
    }

    // sort the list in order of size
    mess.qsort(messcompare);

    while (next_ack < nmsgs 
	   || (next_ack && mess[next_ack-1]->deletes_after_this)
	   || sendmails || safe_deletes.count() > 0)
    {
	if (!isok()) break;
	log(WvLog::Debug4, "next_ack=%s/%s dels=%s sendmails=%s/%s\n",
	    next_ack, nmsgs,
	    mess[0]->deletes_after_this, sendmails,
	    WvSendmailProc::num_sendmails);

	// do any necessary safe deletes
	while (safemode && safe_deletes.count() > 0 &&
		next_req-next_ack < max_requests)
	{
	    cmd(safe_deletes.popstr());
	    if (!response())
	    {
		log(WvLog::Warning, "Failed safe deleting a message!?\n");
		if (flushing)
		    log(WvLog::Warning, "Canceling future deletions "
			"to protect the innocent.\n");
		flushing = false;
		safe_deletes.zap();
	    }

	    if (!isok())
		goto fail;
	}

	while (next_req < nmsgs && next_req-next_ack < max_requests)
	{
	    cmd("retr %s", mess[next_req]->num);
	    next_req++;
	}
    
	MsgInfo *m = mess[next_ack];
	
	while (next_ack > 0 && mess[next_ack-1]->deletes_after_this)
	{
	    if (!response())
	    {
		log(WvLog::Warning, "Failed deleting a message!?\n");
		if (flushing)
		    log(WvLog::Warning, "Canceling future deletions "
			"to protect the innocent.\n");
		flushing = false;
	    }
	   
	    mess[next_ack-1]->deletes_after_this--;
	    
	    if (!isok())
		goto fail;
	}
	
	if (next_ack >= nmsgs)
	{
	    never_select = true;
	    continue_select(20);
	    never_select = false;
	    continue; // only deletions and/or sendmails remaining
	}
	
	if (!response() && isok())
	{
	    next_ack++;
	    log(WvLog::Warning, "Hmm... missing message #%s.\n", m->num);
	    continue;
	}
	
	if (!isok())
	{
	    if (!geterr()) // Keep the older, better?, error message
	        seterr("Aborted.\n");
	    return;
	}
	
	next_ack++;
	
	WvString size;
	if (m->len > 10*1024*1024)
	    size = WvString("%sM", m->len/1024/1024);
	else if (m->len >= 10*1024)
	    size = WvString("%sk", m->len/1024);
	else if (m->len > 1024)
	    size = WvString("%s.%sk", m->len/1024, (m->len*10/1024) % 10);
	else
	    size = WvString("%sb", m->len);
	
	log(WvLog::Debug1, "%-6s %6s ", 
	    WvString("[%s]", mess[next_ack-1]->num), size);
	
	from = "";
	subj = "";
	printed = false;
	in_head = true;
	msgdone = false;
	
	while (WvSendmailProc::num_sendmails >= MAX_PROCESSES && isok())
	{
	    never_select = true;
	    continue_select(1000);
	    never_select = false;
	}

	{
	    if (access(mda, X_OK))
	    {
		log(WvLog::Error, "%s: %s\n", mda, strerror(errno));
		not_found = true;
		goto fail;
	    }
	}
	const char *argv[] = {mda, deliverto, NULL};
	//const char *argv[] = {"dd", "of=/dev/null", NULL};
	WvSendmailProc *p = NULL; 
	if (!explode)
	  {
	    p = new WvSendmailProc(argv, next_ack-1,
				   WvSendmailCallback(this, &WvPopClient::send_done));
	    sendmails++;
	  }
	
	while ((line = blocking_getline(60*1000)) != NULL)
	{
	    if (!isok())
	    {
		if (p)
		  p->kill(SIGTERM);
		seterr("Connection dropped while reading message contents!");
		return;
	    }
	    
	    // remove \r character, if the server gave one
	    cptr = strchr(line, 0) - 1;
	    if (*cptr == '\r')
		*cptr = 0;
	    
	    if (line[0]==0 || line[0]=='\r')
		in_head = false;
	    if (line[0]=='.')
	    {
		if (!strcmp(line, ".") || !strcmp(line, ".\r"))
		{
		    msgdone = true;
		    break; // dot on a line by itself: done this message.
		}
		else
		{
		    // POP servers turn any line *starting with* a dot into
		    // a line starting with dot-dot; so remove the extra
		    // dot.
		    line++;
		}
	    }
	    
	    if (in_head && !strncasecmp(line, "From: ", 6))
	    {
		cptr = strchr(line+6, '<');
		if (cptr)
		{
		    WvString tmp(cptr + 1);
		    cptr = strchr(tmp.edit(), '>');
		    if (cptr)
			*cptr = 0;
		    from = WvString("<%s", tmp);
		}
		else
		    from = line+6;
		trim_string(from.edit());
	    }
	    else if (in_head && explode && !p
		     && !strncasecmp(line, "X-Envelope-To: ", 14))
	    {
		WvString sendto = line+strlen("X-Envelope-To:");
		sendto.edit();
		cptr = strchr(sendto, '@');
		if (cptr)
		    *cptr = 0;
		trim_string(sendto.edit());
		const char *argvE[] = {mda, sendto, NULL};
		p = new WvSendmailProc(argvE, next_ack-1,
				       WvSendmailCallback(this, &WvPopClient::send_done));
		sendmails++;
	    }
	    else if (in_head && !strncasecmp(line, "Subject: ", 9))
	    {
		subj = line+9;
		trim_string(subj.edit());
	    }
	    
	    if (!!from && !!subj && !printed)
	    {
		log(WvLog::Debug1, format, from, subj);
		printed = true;
	    }
	   
	    // ideally we might try to fix the Return-Path here but that
	    // could turn out to be a parsing nightmare so we're going to
	    // just dump it if their POP3 server mangles it
	    //
	    if (p && (!in_head || !ignorerp ||
			strncasecmp(line, "Return-Path: ", 13)))
		p->print("%s\n", line);
	}

	if (isok() && !msgdone)
	{
	    if (p)
	        p->kill(SIGTERM);
	    seterr(ETIMEDOUT);
	    return;
	}

	if (isok() && !printed)
	{
	    if (!subj) subj = "<NO SUBJECT>";
	    if (!from) from = "<NO SENDER";
	    log(WvLog::Debug1, format, from, subj);
	}
	
	if (p)
	{
	    p->done();
	    WvIStreamList::globallist.append(p, true, "sendmail");
	    sendprocs.add(p, false);
	}
    }
    
    cmd("quit");
    if (!response()) goto fail;
    
    close();
    return;
    
fail:
    if (cloned && cloned->geterr())
	seterr("Server connection error: %s", cloned->errstr());
    else if (!cloned || !cloned->isok())
	seterr("Server connection closed unexpectedly (%s bytes left)!",
	       cloned ? inbuf.used() : 0);
    else if (not_found)
	seterr("MDA could not be executed!");
    else
	seterr("Server said something unexpected!");
    return;
}


void WvPopClient::send_done(int count, bool success)
{
    log(WvLog::Debug4, "send_done %s (success=%s)\n", count+1, success);
    
    WvSendmailProc *proc = sendprocs[count];
    if (proc)
	sendprocs.remove(proc);
    
    sendmails--;

    if (!success)
    {
	log(WvLog::Warning, "Error delivering message %s to the MDA.\n",
	    mess[count]->num);
    }
    else
    {
	mess[count]->sent = true;
	
	// remember that we had one more delete message after the most recent
	// request (which might have no relation to the message we're deleting,
	// but we have to count responses carefully!)
	if (flushing)
	{
	    // in safe mode we do all deletes at the end
	    if (safemode)
	    {
	        log(WvLog::Debug3, "Queueing message %s for deletion\n",
			mess[count]->num);
		safe_deletes.append(
			new WvString("dele %s", mess[count]->num), true);
	    }
	    else
	    {
	        cmd("dele %s", mess[count]->num);
	        mess[next_req-1]->deletes_after_this++;
	    }
	}
    }
    
    // wake up the WvPopClient, which may have been waiting for a sendmail
    // to finish.
    never_select = false;
}

